pkg/controller/persistentvolume/persistentvolume_controller.go

```diff
@@ -0,0 +1,1438 @@
+
+// NewPersistentVolumeController creates a new PersistentVolumeController
+func NewPersistentVolumeController(
+	kubeClient clientset.Interface,
+	syncPeriod time.Duration,
+	provisioner volume.ProvisionableVolumePlugin,
+	recyclers []volume.VolumePlugin,
+	cloud cloudprovider.Interface,
+	clusterName string) *PersistentVolumeController {
+
+	broadcaster := record.NewBroadcaster()
+	broadcaster.StartRecordingToSink(&unversioned_core.EventSinkImpl{Interface: kubeClient.Core().Events("")})
+	recorder := broadcaster.NewRecorder(api.EventSource{Component: "persistentvolume-controller"})
+
+	controller := &PersistentVolumeController{
+		kubeClient:                    kubeClient,
+		eventRecorder:                 recorder,
+		runningOperations:             make(map[string]bool),
+		cloud:                         cloud,
+		provisioner:                   provisioner,
+		clusterName:                   clusterName,
+		createProvisionedPVRetryCount: createProvisionedPVRetryCount,
+		createProvisionedPVInterval:   createProvisionedPVInterval,
+	}
+	controller.recyclePluginMgr.InitPlugins(recyclers, controller)
+	if err := controller.provisioner.Init(controller); err != nil {
+		glog.Errorf("PersistentVolumeController: error initializing provisioner plugin: %v", err)
+	}
+
+	volumeSource := &cache.ListWatch{
+		ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+			return kubeClient.Core().PersistentVolumes().List(options)
+		},
+		WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+			return kubeClient.Core().PersistentVolumes().Watch(options)
+		},
+	}
+
+	claimSource := &cache.ListWatch{
+		ListFunc: func(options api.ListOptions) (runtime.Object, error) {
+			return kubeClient.Core().PersistentVolumeClaims(api.NamespaceAll).List(options)
+		},
+		WatchFunc: func(options api.ListOptions) (watch.Interface, error) {
+			return kubeClient.Core().PersistentVolumeClaims(api.NamespaceAll).Watch(options)
+		},
+	}
+
+	controller.initializeController(syncPeriod, volumeSource, claimSource)
+
+	return controller
+}
+
+func (ctrl *PersistentVolumeController) initializeController(syncPeriod time.Duration, volumeSource, claimSource cache.ListerWatcher) {
+	glog.V(5).Infof("Initializing PersistentVolumeController, sync every %s", syncPeriod.String())
+	ctrl.volumes.store, ctrl.volumeController = framework.NewIndexerInformer(
+		volumeSource,
+		&api.PersistentVolume{},
+		syncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    ctrl.addVolume,
+			UpdateFunc: ctrl.updateVolume,
+			DeleteFunc: ctrl.deleteVolume,
+		},
+		cache.Indexers{"accessmodes": accessModesIndexFunc},
+	)
+	ctrl.claims, ctrl.claimController = framework.NewInformer(
+		claimSource,
+		&api.PersistentVolumeClaim{},
+		syncPeriod,
+		framework.ResourceEventHandlerFuncs{
+			AddFunc:    ctrl.addClaim,
+			UpdateFunc: ctrl.updateClaim,
+			DeleteFunc: ctrl.deleteClaim,
+		},
+	)
+}
+
+func (ctrl *PersistentVolumeController) addVolume(obj interface{}) {
```

Comment:
personal note: I finished fixing the PR here. I'll push fixes and go through the rest tomorrow.